Sources/PackageCollections/Storage/PackageCollectionsProfileStorage.swift

```diff
@@ -0,0 +1,339 @@
+        guard buffer.count > 0 else {
+            return .init()
+        }
+        let container = try jsonDecoder.decode(Model.Container.self, from: Data(buffer))
+        return try container.profiles()
+    }
+
+    private func saveToDisk(_ profiles: [PackageCollectionsModel.Profile: [PackageCollectionsModel.PackageCollectionSource]]) throws {
+        if !self.fileSystem.exists(self.path.parentDirectory) {
+            try self.fileSystem.createDirectory(self.path.parentDirectory, recursive: true)
+        }
+        let container = Model.Container(profiles)
+        let buffer = try jsonEncoder.encode(container)
+        try self.fileSystem.writeFileContents(self.path, bytes: ByteString(buffer))
+    }
+
+    private enum Errors: Error {
+        case invalidProfile(PackageCollectionsModel.Profile)
+    }
+}
+
+// MARK: - FilePackageCollectionsProfileStorage Serialization
+
+private enum Model {
+    struct Container: Codable {
+        var data: [String: [Source]]
+
+        init() {
+            self.data = .init()
+        }
+
+        init(_ profiles: [PackageCollectionsModel.Profile: [PackageCollectionsModel.PackageCollectionSource]]) {
+            self.data = .init()
+            profiles.forEach { key, value in
+                self.data[key.profile()] = value.map { $0.source() }
+            }
+        }
+
+        func profiles() throws -> [PackageCollectionsModel.Profile: [PackageCollectionsModel.PackageCollectionSource]] {
+            var profiles = [PackageCollectionsModel.Profile: [PackageCollectionsModel.PackageCollectionSource]]()
+            try self.data.forEach { key, value in
+                try profiles[PackageCollectionsModel.Profile(key)] = value.map { try PackageCollectionsModel.PackageCollectionSource($0) }
+            }
+            return profiles
+        }
+    }
+
+    struct Source: Codable {
+        let type: String
+        let value: String
+    }
+
+    enum SourceType: String {
+        case feed
+    }
+}
+
+// MARK: - Utility
+
+private extension PackageCollectionsModel.Profile {
+    init(_ from: String) {
+        self.init(name: from)
+    }
+
+    func profile() -> String {
+        name
+    }
+}
+
+private extension PackageCollectionsModel.PackageCollectionSource {
+    init(_ from: Model.Source) throws {
+        switch from.type {
+        case Model.SourceType.feed.rawValue:
+            guard let url = URL(string: from.value) else {
+                throw SerializationError.invalidURL(from.value)
+            }
+            self = .feed(url)
+        default:
+            throw SerializationError.unknownType(from.type)
+        }
+    }
+
+    func source() -> Model.Source {
+        switch self {
+        case .feed(let url):
+            return .init(type: Model.SourceType.feed.rawValue, value: url.absoluteString)
+        }
+    }
+}
+
+private enum SerializationError: Error {
+    case unknownType(String)
+    case invalidURL(String)
+}
+
+// FIXME: move to TSC?
```

Comment:
👀

Reply:
Yes, this should almost certainly be in TSC.  I thought the deal with FileLock was also that it was supposed to be available for all file systems, falling back to a Lock if appropriate.

Reply:
removed since not needed any longer